File: src/main/java/me/fzzyhmstrs/amethyst_core/interfaces/Modifiable.java
package me.fzzyhmstrs.amethyst_core.interfaces;


import me.fzzyhmstrs.amethyst_core.modifier_util.AbstractModifierHelper;
import me.fzzyhmstrs.amethyst_core.modifier_util.ModifierInitializer;
import net.minecraft.client.item.TooltipContext;
import net.minecraft.item.ItemStack;
import net.minecraft.text.Text;
import net.minecraft.util.Identifier;

import java.util.Collections;
import java.util.List;

public interface Modifiable {

    default List<Identifier> defaultModifiers(){return Collections.emptyList();}

    default void addModifierTooltip(ItemStack stack, List<Text> tooltip, TooltipContext context){
        AbstractModifierHelper.Companion.getEmptyHelper().addModifierTooltip(stack, tooltip, context);
    }

    default ModifierInitializer getModifierInitializer(){
        return AbstractModifierHelper.Companion.getEmptyHelper();
    }
}

File: src/main/kotlin/me/fzzyhmstrs/amethyst_core/item_util/interfaces/Modifiable.kt
package me.fzzyhmstrs.amethyst_core.item_util.interfaces

import me.fzzyhmstrs.amethyst_core.coding_util.AcText
import me.fzzyhmstrs.amethyst_core.coding_util.Addable
import me.fzzyhmstrs.amethyst_core.modifier_util.AbstractModifier
import me.fzzyhmstrs.amethyst_core.modifier_util.AbstractModifierHelper
import me.fzzyhmstrs.amethyst_core.modifier_util.ModifierHelper
import net.minecraft.item.ItemStack
import net.minecraft.text.MutableText
import net.minecraft.text.Text
import net.minecraft.util.Formatting
import net.minecraft.util.Identifier

/**
 * An item or object that extends Modifiable is intended to mesh with the Modifier system. See the [ModifiableScepterItem][me.fzzyhmstrs.amethyst_core.item_util.ModifiableScepterItem] for a default implementation.
 *
 * Description is left intentionally vague as there are any number of directions one can take the Modifier system.
 */

interface Modifiable<T:AbstractModifier<T>> {

    val defaultModifiers: MutableList<Identifier>

    fun getModifierHelper(): AbstractModifierHelper<T>
}